cl/api/utils.py

```diff
@@ -346,6 +349,148 @@ def allow_request(self, request, view):
         return self.throttle_success()
 
 
+class CitationCountRateThrottle(ExceptionalUserRateThrottle):
+    """
+    Limits the rate of API calls that may be made by a users based on the
+    number of citations they try to look up.
+    """
+
+    def get_citation_count_from_request(self, request, view) -> int:
+        """
+        Gets the number of citations from a request.
+
+        This helper method retrieves the number of citations from a request
+        object. It first validates the data using the `validate_request_data`
+        method. If valid, it extracts the citations list and it returns the
+        number of citations in the list.
+
+        Returns:
+            int: The number of citations as an integer.
+        """
+        validated_data = view.validate_request_data(request)
+        text = validated_data.get("text", None)
+        if not text:
+            # Since the 'text' key is missing from the request, the user is
+            # likely trying to retrieve opinions using a reporter, volume,
+            # and page combination. This approach allows looking one citation
+            # at a time.
+            return 1
+
+        citation_objs = filter_out_non_case_law_and_non_valid_citations(
+            eyecite.get_citations(text)
+        )
```

Comment:
I don't suppose there's a way to do this in the request and pass it to the throttle, so we only have to do it once? I think not, but I thought I'd ask.

Reply:
Actually, I believe there's a way to compute this list of citations just once. Since the `get_citation_count_from_request` method has access to both the request object and the view object, we can store the list of citations as an attribute of the view instance. This way, we can avoid redundant computations whenever a request is allowed.

The new `get_citation_count_from_request` helper will look like this: 

```python
def get_citation_count_from_request(self, request, view) -> int:
    ...
    citation_objs = filter_out_non_case_law_and_non_valid_citations(
         eyecite.get_citations(text)
    )
    view.citations = citations_objs
    return len(citations_objs)
```


Then, in the API view, we'll apply the following tweaks:


```diff
def create(self, request: Request, *args, **kwargs):
   ...
   if text:
-       citation_objs = eyecite.get_citations(text)
-       citation_objs = filter_out_non_case_law_and_non_valid_citations(
-               citation_objs
-       )
-       if not citation_objs:
+       if not self.citations:
            return Response([])
         
-       for idx, citation in enumerate(citation_objs):
+.      for idx, citation in enumerate(self.citations):
            ...

```

Let me know what you think

Reply:
I'd be confused how the view object got a `citations` attribute, but otherwise it makes sense. Maybe just add a comment in there:

```
# self.citations is set by the throttle class. This is a performance
# enhancement to avoid parsing the citations twice, first in the 
# throttle and then again here.
if not self.citations:
``` 

Works for me?

Reply:
> Maybe just add a comment in there

great! I'll do that. Thanks 👍